### PR TITLE
8272905: Consolidate discovered lists processing

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple change of merging discovered list processing for four kinds of non-strong refs and some general cleanup around.

Test: tier1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272905](https://bugs.openjdk.java.net/browse/JDK-8272905): Consolidate discovered lists processing

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to cfcbad70efeee223fac22634eaea7607d04c3cb0
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 3650b5d60bf877b69831f6ab2352e3d7d8452b06

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5242/head:pull/5242` \
`$ git checkout pull/5242`

Update a local copy of the PR: \
`$ git checkout pull/5242` \
`$ git pull https://git.openjdk.java.net/jdk pull/5242/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5242`

View PR using the GUI difftool: \
`$ git pr show -t 5242`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5242.diff">https://git.openjdk.java.net/jdk/pull/5242.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shared/referenceProcessor.cpp [hunk 1]
```diff
@@ -367,44 +366,12 @@ size_t ReferenceProcessor::process_final_keep_alive_work(DiscoveredList& refs_li
...
-                                                     OopClosure*        keep_alive,
-                                                     VoidClosure*       complete_gc) {
-  DiscoveredListIterator iter(refs_list, keep_alive, is_alive);
-  while (iter.has_next()) {
-    iter.load_ptrs(DEBUG_ONLY(!discovery_is_atomic() /* allow_null_referent */));
-
-    oop const referent = iter.referent();
-
-    if (referent == NULL || iter.is_referent_alive()) {
-      iter.make_referent_alive();
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/referenceProcessor.cpp:389]: The handling for `NULL` referents is (just a bit) different for phantom refs than for the others, but seems inconsequential.

**Review by tschatzl — APPROVED** [review]
Looks good to me as far as I can tell apart from the minor suggestion in `RefProcTask::process_discovered_list`, but I'm not too confident about the ins and outs of this code.

File: src/hotspot/share/gc/shared/referenceProcessor.cpp [hunk 2]
```diff
@@ -451,21 +418,47 @@ size_t ReferenceProcessor::total_reference_count(ReferenceType type) const {
...
-                  DiscoveredList list[],
-                  BoolObjectClosure* is_alive,
-                  OopClosure* keep_alive,
-                  bool do_enqueue_and_clear,
-                  ReferenceType ref_type) {
-    size_t const removed = _ref_processor.process_soft_weak_final_refs_work(list[worker_id],
-                                                                            is_alive,
-                                                                            keep_alive,
-                                                                            do_enqueue_and_clear);
+  // Only Final refs are not enqueued.
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/referenceProcessor.cpp:448]: s/enqueued./enqueued and cleared here./

File: src/hotspot/share/gc/shared/referenceProcessor.cpp [hunk 3]
```diff
@@ -451,21 +418,47 @@ size_t ReferenceProcessor::total_reference_count(ReferenceType type) const {
   return total_count(list);
 }
 
+void RefProcTask::process_discovered_list(uint worker_id,
+                                          ReferenceType ref_type,
+                                          BoolObjectClosure* is_alive,
+                                          OopClosure* keep_alive) {
+  ReferenceProcessor::RefProcSubPhases subphase;
+  DiscoveredList* dl;
+  switch(ref_type) {
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/referenceProcessor.cpp:427]: Missing space after "switch" per style guide.

**Review by kimbarrett — APPROVED** [review]
Looks good.  Just a couple very minor nits.  I re-examined the discussion and change for JDK-8202845 as part of examining this.

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**